packages/flutter_tools/lib/src/dart/pub.dart

```diff
@@ -557,12 +572,18 @@ class _DefaultPub implements Pub {
       final Directory globalDirectoryPub = _fileSystem.directory(
         _fileSystem.path.join(globalDirectory!.path, 'hosted', 'pub.dartlang.org')
       );
-      joinCaches(
+      final String? failedDirectory = joinCaches(
         fileSystem: _fileSystem,
         globalCachePath: globalDirectoryPub.path,
         localCachePath: localDirectoryPub.path,
       );
-      _fileSystem.directory(localCachePath).deleteSync(recursive: true);
+      if (failedDirectory != null) {
```

Comment:
Maybe it would be better if the deletion happened inside of joinCaches. This feels like hte logic is being spread between the two functions.